### PR TITLE
Adding API Key generation implementation

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Purpose
Fix for: https://github.com/wso2/apk/issues/335

## Curl

curl --location --request POST 'http://localhost:80/api/am/devportal/applications/8cc55d57-12aa-4238-8a21-7932b6c2bab2/api-keys/PRODUCTION/generate' \
--header 'Content-Type: application/json' \
--header 'Cookie: route=1671001178.544.722.749029|ba569ae14c8227f9364b64819e5124d9' \
--data-raw '{
  "validityPeriod": -1,
  "additionalProperties": {
    "permittedIP": "",
    "permittedReferer": ""
  }
}'

## Response

{
    "apikey": "eyJhbGciOiJSUzI1NiIsICJ0eXAiOiJKV1QiLCAia2lkIjoiMDFlZDdjMzgtNTg0Zi0xZjAwLWFkMGEtMTg0MzRiY2JkZGQ0In0.eyJpc3MiOiJodHRwczovL2xvY2FsaG9zdDo5NDQzL29hdXRoMi90b2tlbiIsICJzdWIiOiIiLCAiYXVkIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6OTQ0My9vYXV0aDIvdG9rZW4iLCAiZXhwIjoxNjcxMDg1MzY3LCAibmJmIjoxNjcxMDgxNzY3LCAiaWF0IjoxNjcxMDgxNzY3LCAianRpIjoiMDFlZDdjMzgtODJhMS0xYWRlLThmYTgtYmY4YThjM2VkNTRmIiwgImtleXR5cGUiOiJQUk9EVUNUSU9OIiwgInBlcm1pdHRlZFJlZmVyZXIiOiIiLCAicGVybWl0dGVkSXAiOiIiLCAidG9rZW5fdHlwZSI6IkFQSUtleSIsICJ0aWVySW5mbyI6IiIsICJzdWJzY3JpYmVkQVBJcyI6W3sibmFtZSI6Imh0dHAtYmluLWFwaSIsICJjb250ZXh0IjoiL2h0dHAtYmluLWFwaS8xLjAuOCIsICJ2ZXJzaW9uIjoiMS4wLjgiLCAicHVibGlzaGVyIjoiYXBrdXNlciIsICJ1dWlkIjoiNGMzY2RhNGQtMjRhNC00MjQ4LWI4MjQtOWIwMDAzZTlhNTExIn1dLCAiYXBwbGljYXRpb24iOnsibmFtZSI6IlNhbXBsZUFwcDEiLCAidXVpZCI6IjhjYzU1ZDU3LTEyYWEtNDIzOC04YTIxLTc5MzJiNmMyYmFiMiIsICJpZCI6IjEiLCAidGllciI6IjYwUGVyTWluIiwgInRpZXJRdW90YVR5cGUiOiIifX0.Uhw13EmIb94BpEKBIQjSgBlCeYNlcnLS0sL4yWGe3ULBMrRRjdEiQ3N-4ibxqyoaXC2b0Cx8xla5VYOFt_-mOwI66i_l_eL4R1L84Izlwt3pMTzXfrFwE6dUGb8Wz59VTrLXhm1jFw-Sj2RfRKFlpEPoWjBKzl8rJcVRsyt0TAter-gu-7UOfQaucoMiPcxUH_lULg7nSVoefwC-gkshE-9Dpplb_JGFmEQ5WOqqfuPcKi5swuZLl6GFWyko0eW8z8Q4huonCpWA9g4rsnv1NknKa8YzAFlJWCYKlfCkd0o34Dc_-rLELeFoAqpH50tsz32ry63YohPSQVJPzRruBg",
    "validityTime": 3600
}

### CODE REVIEW

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 1]
```diff
@@ -86,3 +86,72 @@ function deleteApplication(string appId, string organization) returns string|err
...
+                validityPeriod = payloadValPeriod;
+            } 
+            record {} addProperties = {};
+            record {}? payloadAddProperties = payload.additionalProperties;
+            if payloadAddProperties is record {} {
+                addProperties = payloadAddProperties;
+            }
+
+            // retrieve subscribed APIs
+            string?|SubscriptionList|error subscriptions =  getSubscriptions(null,appId,null,0,0,org);
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/applicationImpl.bal:109]: I suppose we have formatting issues in here. shall we do a `bal format` for these changes?
```suggestion
            string?|SubscriptionList|error subscriptions =  getSubscriptions(null, appId, null, 0, 0, org);
```

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 2]
```diff
@@ -86,3 +86,72 @@ function deleteApplication(string appId, string organization) returns string|err
...
+    string?|Application|error application = getApplicationById(appId, org);
+    if application !is Application {
+        return error("Invalid Application. Application with application id:" + appId + " not found");
+    } else {
+        boolean userAllowed = checkUserAccessAllowedForApplication(application, user);
+        if userAllowed {
+            int validityPeriod = 0;
+            int? payloadValPeriod = payload.validityPeriod;
+            if payloadValPeriod is int {
+                validityPeriod = payloadValPeriod;
```
[the remark this answers is not in the document]
Comment (AmaliMatharaarachchi) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal:100]: please handle errors for other payload parsing issues below.
e.g. https://github.com/wso2/apk/pull/359/files#diff-a41c513961f1989b587eb76d0c876656832c7ea78a892d72a3bf6d9d2911c43fR127
https://github.com/wso2/apk/pull/359/files#diff-a41c513961f1989b587eb76d0c876656832c7ea78a892d72a3bf6d9d2911c43fR121
https://github.com/wso2/apk/pull/359/files#diff-a41c513961f1989b587eb76d0c876656832c7ea78a892d72a3bf6d9d2911c43fR104

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 3]
```diff
@@ -86,3 +86,72 @@ function deleteApplication(string appId, string organization) returns string|err
...
+                if subArray is Subscription[] {
+                    foreach Subscription item in subArray {
+                        string? apiUUID = item.apiId;
+                        if apiUUID is string {
+                            string?|API|error api = getAPIByAPIId(apiUUID,org);
+                            if api is API {
+                                apiList.push(api);
+                            }
+                        } else {
+                            log:printDebug(apiUUID.toString());
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/applicationImpl.bal:122]: Is this intentional?

File: devportal/devportal-domain-service/ballerina/JWTGenerator.bal [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    //int? id = app.id;
+    string? owner = app.owner;
+    string? tier = app.throttlingPolicy;
+    application["name"] = app.name;
+    if uuid is string{
+        application["uuid"] = uuid;
+    }
+    // if id is int {
+    //     application["id"] = id;
+    // }
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/JWTGenerator.bal:96]: Is this intentional?

File: devportal/devportal-domain-service/ballerina/jwtTokenInfo.bal [hunk 5]
```diff
@@ -0,0 +1,9 @@
+public type JWTTokenInfo record {
+    Application application;
+    API[] subscribedAPIs;
+    string subscriber;
+    string expireTime;
+    string keyType;
+    string permittedIP;
+    string permittedReferrer;
+};
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/jwtTokenInfo.bal]: fixed

File: devportal/devportal-domain-service/ballerina/devportal-api_service.bal [hunk 6]
```diff
@@ -196,8 +196,16 @@ service /api/am/devportal on ep0 {
     // }
     // resource function post applications/[string applicationId]/'oauth\-keys/[string keyMappingId]/'generate\-token(@http:Header string? 'if\-match, @http:Payload ApplicationTokenGenerateRequest payload) returns ApplicationToken|BadRequestError|NotFoundError|PreconditionFailedError {
     // }
-    // resource function post applications/[string applicationId]/'api\-keys/[string keyType]/generate(@http:Header string? 'if\-match, @http:Payload APIKeyGenerateRequest payload) returns APIKey|BadRequestError|NotFoundError|PreconditionFailedError {
-    // }
+    resource function post applications/[string applicationId]/'api\-keys/[string keyType]/generate(@http:Header string? 'if\-match, @http:Payload APIKeyGenerateRequest payload) returns APIKey|BadRequestError|NotFoundError|PreconditionFailedError|InternalServerErrorError|error {
+        APIKey|error apiKey = check generateAPIKey(payload, applicationId, keyType, "apkuser", "carbon.super");
+        if apiKey is APIKey {
+            log:printDebug(apiKey.toString());
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/devportal-api_service.bal]: removed and fixed

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 7]
```diff
@@ -86,3 +86,72 @@ function deleteApplication(string appId, string organization) returns string|err
...
+        return apiKey;
+    } else {
+        log:printDebug(token.message());
+        return error("Error while generating token");
+    }
+}
+
+function checkUserAccessAllowedForApplication(Application application, string user) returns boolean {
+    return true;
+}
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal]: fixed

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 8]
```diff
@@ -86,3 +86,72 @@ function deleteApplication(string appId, string organization) returns string|err
...
+                    }
+                }
+            } else if subscriptions is error {
+                log:printDebug(subscriptions.message());
+            }
+
+            APIKey|error apiKey = generateAPIKeyForApplication(user, application, apiList, keyType, validityPeriod, addProperties);
+            return apiKey;
+
+
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal]: fixed

File: devportal/devportal-domain-service/ballerina/jwtTokenInfo.bal [hunk 9]
```diff
@@ -0,0 +1,9 @@
+public type JWTTokenInfo record {
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/jwtTokenInfo.bal:19]: fixed

File: devportal/devportal-domain-service/ballerina/init.bal [hunk 10]
```diff
@@ -19,17 +19,31 @@
 import ballerina/log;
 import ballerinax/java.jdbc;
 import ballerina/sql;
+import ballerina/uuid;
 
 configurable DatasourceConfiguration datasourceConfiguration = ?;
 configurable ThrottlingConfiguration throttleConfig = ?;
+string kid = uuid:createType1AsString();
```
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/init.bal:26]: used and fixed

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal (hunk 2 above)
[the remark this answers is not in the document]
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal:100]: fixed all places

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal (hunk 1 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal]: fixed

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal (hunk 3 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal]: fixed

File: devportal/devportal-domain-service/ballerina/JWTGenerator.bal (hunk 4 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/JWTGenerator.bal]: fixed

**Review by AmaliMatharaarachchi — APPROVED** [review]

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 11]
```diff
@@ -86,3 +86,74 @@ function deleteApplication(string appId, string organization) returns string|err
...
+                            if api is API {
+                                apiList.push(api);
+                            }
+                        } else {
+                            log:printDebug("Invalid API UUID found:" + apiUUID.toString());
+                        }
+                    }
+                }
+            } else if subscriptions is error {
+                log:printDebug(subscriptions.message());
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/applicationImpl.bal:131]: shall we handle error

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal [hunk 12]
```diff
@@ -86,3 +86,74 @@ function deleteApplication(string appId, string organization) returns string|err
...
+                if subArray is Subscription[] {
+                    foreach Subscription item in subArray {
+                        string? apiUUID = item.apiId;
+                        if apiUUID is string {
+                            string?|API|error api = getAPIByAPIId(apiUUID, org);
+                            if api is API {
+                                apiList.push(api);
+                            }
+                        } else {
+                            log:printDebug("Invalid API UUID found:" + apiUUID.toString());
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/applicationImpl.bal:126]: shall we handle this error

File: devportal/devportal-domain-service/ballerina/config.bal [hunk 13]
```diff
@@ -49,4 +52,23 @@ type DatasourceConfiguration record {
...
     DatasourceConfiguration datasourceConfiguration;
+    TokenIssuerConfiguration tokenIssuerConfiguration;
+    KeyStores keyStores;
 };
+
+public type TokenIssuerConfiguration record {|
+    string issuer = "https://apim.wso2.com/oauth2/token";
+    string audience = "https://apim.wso2.com/oauth2/token";
+    string keyId = "gateway_certificate_alias";
+    decimal expTime = 3600;
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/ballerina/config.bal:63]: isn't this should be configurable

File: devportal/devportal-domain-service/docker/devportal/conf/Config.toml [hunk 14]
```diff
@@ -6,5 +6,20 @@
...
+    [wso2.devportal_service.keyStores.tls]
+    path = "/home/wso2apk/devportal/security/wso2carbon.key"
+    [wso2.devportal_service.keyStores.signing]
+    path = "/home/wso2apk/devportal/security/wso2carbon.key"
+
+    [wso2.devportal_service.issuerConfig]
+    issuer = "https://apim.wso2.com/oauth2/token"
+    audience = "https://apim.wso2.com/oauth2/token"
+    keyId = "gateway_certificate_alias"
+    expTime = 3600.0
```
Comment (AmaliMatharaarachchi) [inline devportal/devportal-domain-service/docker/devportal/conf/Config.toml:25]:
```suggestion
    expTime = 3600.0
    
```

File: devportal/devportal-domain-service/ballerina/config.bal (hunk 13 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/config.bal:63]: yes, this provided through configurations these are default values

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal (hunk 11 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal:131]: resolved

File: devportal/devportal-domain-service/ballerina/applicationImpl.bal (hunk 12 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/ballerina/applicationImpl.bal:126]: resolved

File: devportal/devportal-domain-service/docker/devportal/conf/Config.toml (hunk 14 above)
Comment (CrowleyRajapakse) [reply devportal/devportal-domain-service/docker/devportal/conf/Config.toml:25]: resolved

**Pull request merged**